xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1157 @@
+/**
+ * xrdp: A Remote Desktop Protocol server.
+ *
+ * Copyright (C) Jay Sorg 2019
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * MS-RDPEGFX
+ */
+
+#if defined(HAVE_CONFIG_H)
+#include <config_ac.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "arch.h"
+#include "os_calls.h"
+#include "parse.h"
+#include "xrdp.h"
+#include "xrdp_egfx.h"
+#include "libxrdp.h"
+#include "xrdp_channel.h"
+#include <limits.h>
+
+#define MAX_PART_SIZE 0xFFFF
+
+/******************************************************************************/
+int
+xrdp_egfx_send_data(struct xrdp_egfx *egfx, const char *data, int bytes)
+{
+    int error;
+    int to_send;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_send_data:");
+
+    if (bytes <= 1500)
+    {
+        error = libxrdp_drdynvc_data(egfx->session, egfx->channel_id,
+                                     data, bytes);
+    }
+    else
+    {
+        error = libxrdp_drdynvc_data_first(egfx->session, egfx->channel_id,
+                                           data, 1500, bytes);
+        data += 1500;
+        bytes -= 1500;
+        while ((bytes > 0) && (error == 0))
+        {
+            to_send = bytes;
+            if (to_send > 1500)
+            {
+                to_send = 1500;
+            }
+            error = libxrdp_drdynvc_data(egfx->session, egfx->channel_id,
+                                         data, to_send);
+            data += to_send;
+            bytes -= to_send;
+        }
+    }
+    return error;
+}
+
+/******************************************************************************/
+int
+xrdp_egfx_send_s(struct xrdp_egfx *egfx, struct stream *s)
+{
+    int error;
+    int bytes;
+
+    if (s == NULL)
+    {
+        return 1;
+    }
+    bytes = (int) (s->end - s->data);
+    error = xrdp_egfx_send_data(egfx, s->data, bytes);
+    return error;
+}
+
+/******************************************************************************/
+struct stream *
+xrdp_egfx_create_surface(struct xrdp_egfx_bulk *bulk, int surface_id,
+                         int width, int height, int pixel_format)
+{
+    int bytes;
+    struct stream *s;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_create_surface:");
+    make_stream(s);
+    init_stream(s, 8192);
+    /* RDP_SEGMENTED_DATA */
+    out_uint8(s, 0xE0); /* descriptor = SINGLE */
+    /* RDP8_BULK_ENCODED_DATA */
+    out_uint8(s, 0x04); /* header = PACKET_COMPR_TYPE_RDP8 */
```

Comment:
Small suggestion (not an issue), you could use a `define` for the `PACKET_COMPR_TYPE_RDP8`.

Reply:
Done!